Tools/Scripts/webkitpy/benchmark_runner/http_server_driver/http_server/base_http_server.py

```diff
@@ -0,0 +1,96 @@
+#!/usr/bin/env python3
+
+from http.server import SimpleHTTPRequestHandler, HTTPServer
+import argparse
+import logging
+import os
+import sys
+import socket
+import threading
+# Since we execute this script directly as a subprocess, we need to ensure
+# that Tools/Scripts is in sys.path for the next imports to work correctly.
+script_dir = os.path.abspath(os.path.join(
+    os.path.dirname(os.path.abspath(__file__)), '../../../..'))
+if script_dir not in sys.path:
+    sys.path.insert(0, script_dir)
+
+
+_log = logging.getLogger(__name__)
+
+
+class Handler(SimpleHTTPRequestHandler):
+    web_root = None
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs, directory=self.web_root)
+
+    def log_message(self, format, *args):
+        _log.info(format % args)
+
+    def do_GET(self):
+        if self.path.startswith('/shutdown'):
+            self.send_response(200)
+            self.end_headers()
+            try:
+                shutdown_thread = threading.Thread(target=self.server.shutdown)
+                shutdown_thread.start()
+                sys.exit(0)
+            except Exception as e:
+                _log.error('Error shutting down HTTP server: %s' % e)
+                sys.exit(1)
+
+        else:
+            super().do_GET()
+
+    def do_POST(self):
+        if self.path.startswith('/report'):
+            length = int(self.headers['content-length'])
+            try:
+                content = self.rfile.read(length)
+                sys.stdout.write(content.decode('utf-8'))
+                sys.stdout.flush()
+                self.send_response(200)
+            except Exception as e:
+                _log.error('Error reading and writing report: %s' % e)
+                self.send_response(500)
+            self.end_headers()
+            try:
+                shutdown_thread = threading.Thread(target=self.server.shutdown)
```

Comment:
Two questions:
1. this seems repeating, may worth a helper function.
2. Curious, is this the recommended way for a graceful shutdown? What will happen if we just do `sys.exit(0)`?

Reply:
Using self.server.shutdown() is the recommended way of shutting down http.server gracefully, this is running on a separate thread in the case there are other incoming requests needing to be handled before the shutdown. 

sys.exit(0) works and could be a little bit faster, but this could possibly cause unexpected behavior in the future.

I'll be adding a helper function to handle the server shutdown.